src/server/stream_family.cc

```diff
@@ -1192,97 +1240,245 @@ void StreamFamily::XRevRange(CmdArgList args, ConnectionContext* cntx) {
+
   auto cb = [&](Transaction* t, EngineShard* shard) {
     ShardId sid = shard->shard_id();
-    xread_resp[sid] = OpRead(t->GetOpArgs(shard), t->GetShardArgs(shard->shard_id()), read_opts);
+    last_ids_res[sid] = OpLastIDs(t->GetOpArgs(shard), t->GetShardArgs(shard->shard_id()));
     return OpStatus::OK;
   };
-  OpStatus result = cntx->transaction->ScheduleSingleHop(std::move(cb));
-  CHECK_EQ(OpStatus::OK, result);
+  trans->Execute(std::move(cb), false);
+
+  unordered_map<string_view, streamID> last_ids;
+  for (auto res : last_ids_res) {
+    if (!res) {
+      return res.status();
+    }
+
+    for (auto& e : *res) {
+      last_ids.emplace(e.first, e.second);
+    }
+  }
+  return last_ids;
+}
+
+void StreamFamily::XRead(CmdArgList args, ConnectionContext* cntx) {
+  auto opts = ParseReadArgsOrReply(args, cntx);
+  if (!opts) {
+    return;
+  }
+
+  cntx->transaction->Schedule();
+
+  auto last_ids = StreamLastIDs(cntx->transaction);
+  if (!last_ids) {
+    // Close the transaction.
+    auto close_cb = [&](Transaction* t, EngineShard* shard) { return OpStatus::OK; };
+    cntx->transaction->Execute(std::move(close_cb), true);
+
+    if (last_ids.status() == OpStatus::WRONG_TYPE) {
+      (*cntx)->SendError(kWrongTypeErr);
+      return;
+    }
+
+    return (*cntx)->SendNullArray();
+  }
+
+  // Resolve '$' IDs and check if there are any streams with entries that can
+  // be resolved without blocking.
+  bool block = true;
+  for (auto& [stream, requested_id] : opts->stream_ids) {
+    if (auto last_id_it = last_ids->find(stream); last_id_it != last_ids->end()) {
+      streamID last_id = last_id_it->second;
+
+      // Resolve $ to the last ID in the stream.
+      if (requested_id.last_id) {
+        requested_id.val = last_id;
+        // We only include messages with IDs greater than the last message so
+        // increment the ID.
+        streamIncrID(&requested_id.val);
+        requested_id.last_id = false;
+        continue;
+      }
+
+      if (streamCompareID(&last_id, &requested_id.val) >= 0) {
+        block = false;
+      }
+    }
+  }
+
+  if (block) {
+    // If BLOCK is not set just return an empty array as there are no resolvable
+    // entries.
+    if (opts->timeout == -1 || cntx->transaction->IsMulti()) {
+      // Close the transaction and release locks.
+      auto close_cb = [&](Transaction* t, EngineShard* shard) { return OpStatus::OK; };
+      cntx->transaction->Execute(std::move(close_cb), true);
+      return (*cntx)->SendNullArray();
+    }
+
+    auto wcb = [](Transaction* t, EngineShard* shard) {
+      return t->GetShardArgs(shard->shard_id());
+    };
+
+    auto tp = (opts->timeout) ? chrono::steady_clock::now() + chrono::milliseconds(opts->timeout)
+                              : Transaction::time_point::max();
+
+    auto* stats = ServerState::tl_connection_stats();
+    ++stats->num_blocked_clients;
+    bool wait_succeeded = cntx->transaction->WaitOnWatch(tp, std::move(wcb));
```

Comment:
`++stats->num_blocked_clients` wrapping code is duplicated 3 times now. Time to move it into WaitOnWatch

Reply:
sure updated